### ISSUE TITLE
Bad Credentials

### ISSUE THREAD

**Issue by lm12000** [hrkfdn/ncspot#1500]
**Describe the bug**
I can't login with my Spotify Family Account.
I am assuming it has to do with the Account being a shared one in a Family Plan ? (Which is Premium.)

**To Reproduce**
1. Try Login.
2. "Connection error: Login failed with reason: Bad Credentials"

**Expected behavior**
Being able to use the application.

**System (please complete the following information):**
 - OS: Linux Pop!_OS 22.04 LTS
 - Installed from: flatpak

PS: I am really new to Linux so please forgive me if this is just a simple fix or something.
I did try the actual username of the account as well as the email linked to it. Both did not work.

**Comment by Apollo-XIV** [comment]
having the same issue here, also on a family account. It's worked for me in the past but now doesn't, although I'm unsure of when that changed. I've checked multiple times and my password works through the browser and I'm confident I'm entering it correctly.

**Comment by hrkfdn** [comment]
It's probably related to this: https://github.com/librespot-org/librespot/issues/1308

Could be we need to change our login method

**Comment by lorenzMehring** [comment]
> having the same issue here, also on a family account. It's worked for me in the past but now doesn't, although I'm unsure of when that changed. I've checked multiple times and my password works through the browser and I'm confident I'm entering it correctly.

This is exactly my problem as well. 
I was now somehow logged out of my spotify on my phone etc. too, which is strange.
When I tried logging back in my password wasnt working anymore.
I could eventually log back in over a link through email from spotify.
Now I changed the password and it still does not work. Weird.

**Comment by lorenzMehring** [comment]
> It's probably related to this: [librespot-org/librespot#1308](https://github.com/librespot-org/librespot/issues/1308)
> 
> Could be we need to change our login method

Ah I see, so its a bigger problem with the login method entirely, not just a Family Plan thing?

**Comment by signaleleven** [comment]
Seems like it was rolled out for different users at different times. I have a partner plan and I was also affected today (and Spotify Connect Home assistant addon, which uses librespot like ncspot - retried so much I got forced to change my password) whereas in the librespot issue it looks like it's being going on for over 2 months for other users.

I would follow the issue linked above :(

**Comment by cyqsimon** [comment]
Affected me today on a family plan. Seems like Spotify broke their own API.

**Comment by geekmuse** [comment]
Regular premium user here (no Family plan) and am experiencing the same issue.

**Comment by jpringle1** [comment]
Installed ncspot for the first time today and ran into this issue. I assume i was doing something wrong, but it seems it might just be very unfortunate timing that this bug has cropped up just a few days before i tried to use this :/

**Comment by donwlewis** [comment]
Yup.  I also ran into this issue.

**Comment by psaumur** [comment]
Am using a Family Plan and had ncspot working on Debian.

When I tried authenticating manually on Mint Edge, it failed a few times, until I copied over the credentials.json file from the working Debian instance.

After listening for a while, Spotify decided to reset my password so my credentials were no longer good.

Trying to reset my credentials for both Debian and Mint Edge have failed today.

**Comment by RobinCamarasa** [comment]
Same issue here using a Duo plan

**Comment by tomasApo** [comment]
Same issue here with family plan

**Comment by psaumur** [comment]
> As others have already mentioned, the credentials generated via the zeroconf authentication mechanism appear long-lived and is the officially sanctioned way to authenticate Spotify Connect devices. I noted some people using downstream clients (those depending upon librespot) were having difficulty in generating the credentials.json file, so I [thinly wrapped just this functionality]([(https://github.com/dspearson/librespot-auth])) and tried to make it as trivial to use as I could think; maybe it helps others who stumble here with similar problems.

[(https://github.com/dspearson/librespot-auth])

Downloading the Linux Spotify client, running Dominic's Rust/Cargo application (link above) in the background, pulls out working credentials. Everything works again.

**Comment by psaumur** [comment]
> I tried this and got the credentials.json but after putting it in the .cache/ncspot/librespot as well as the .config/ncspot/credentials I still get prompted to put in my credentials at the login screen. Am I doing this wrong?

Credentials, for ncspot, by default, go:

~/.var/app/io.github.hrkfdn.ncspot/cache/ncspot/librespot/

or wherever the "cache" path is when you type `ncspot info` + /librespot/

**Comment by chocoblin** [comment]
> > As others have already mentioned, the credentials generated via the zeroconf authentication mechanism appear long-lived and is the officially sanctioned way to authenticate Spotify Connect devices. I noted some people using downstream clients (those depending upon librespot) were having difficulty in generating the credentials.json file, so I [thinly wrapped just this functionality](https://github.com/dspearson/librespot-auth) and tried to make it as trivial to use as I could think; maybe it helps others who stumble here with similar problems.
> 
> Downloading the Linux Spotify client, running Dominic's Rust/Cargo application (link above) in the background, pulls out working credentials. Everything works again.

can you please write out the steps for a noob like me to follow? I don;t understand what to do to install librespot-auth even after going to that page

**Comment by cyqsimon** [comment]
@chocoblin `git clone` then `cargo run`. After that follow README.

**Comment by Zardexd** [comment]
Having this issue with app/dev.alextren.Spot/x86_64/stable flatpak too , just logged with same credentials from browser.

**Comment by thecisco** [comment]
@cyqsimon What exactly is git clone and cargo run? I just got into Linux and got the flatpak ver. of ncspot and I am having the same issue. Sorry for not being knowledgeable in these topics, that is why I must inquire.

**Comment by Zardexd** [comment]
> @cyqsimon What exactly is git clone and cargo run? I just got into Linux and got the flatpak ver. of ncspot and I am having the same issue. Sorry for not being knowledgeable in these topics, that is why I must inquire.

TLDR - I'm almost certain, nothing is wrong on your side. It's likely Spotify devs just changed something in API's internals, and you can't just send your credentials via API, you need an API key. 

Sources, and why I think that? 

I went thru a lot of spotify clients and their issue pages. All have this bad credentials issue. For exception of Spotube , they're using web-page based log in. 

Answers to questions. 

Git clone is a command in Git that allows you to clone/download someone's repository. (usually it used for storing managing, collaboration and version control of source code) 

Cargo is a build system and packet manager for Rust programming language, it allows you to compile (assemble program from its source code) the Rust code and manage packages.

**Comment by cyqsimon** [comment]
> @cyqsimon What exactly is git clone and cargo run? I just got into Linux and got the flatpak ver. of ncspot and I am having the same issue. Sorry for not being knowledgeable in these topics, that is why I must inquire.

Use Google first please 🤦.

**Comment by cyqsimon** [comment]
Not trying to be condescending, but questions like these waste my time and yours. Mine because I could be doing something else; yours because there are plenty of answers online that do a better job than I ever could.

**Comment by Zardexd** [comment]
> What exactly is git clone and cargo run? I just got into Linux and got the flatpak ver. of ncspot and I am having the same issue. Sorry for not being knowledgeable in these topics, that is why I must inquire.

Allows to clone (download) git repository. Repo's are used to store,  collaborate and manage code changes. 

Cargo is a package and build system for Rust programming language. Allows you to assemble a program from its sources (usually stored in repo's), or install packets/libraries that are used throughout the process of developing software. 

Don't be afraid to google things. Most of questions were already answered.

**Comment by roblox-model** [comment]
For some reason librespot-auth's device doesnt appear / doesn't get recognized in the device list at all so i can't generate a credentials.json, could anyone send me a template on what the credentials.json contain and i'll try adding the stuff by myself please ?

**Comment by psaumur** [comment]
If you are not seeing it, it's prob. due to trying to run Spotify in a
browser - instead of running the application directly in the Linux
environment.
I ran into this same issue and once I was running it locally, while running
the "device" to pull credentials, it dumps the credentials out into the
.json file

**Comment by roblox-model** [comment]
well i've tried running the application first, tried using the browser and even the app on my phone to see if airplay detected it but nothing appears

**Comment by Zocker1999NET** [comment]
I seem not able to fully resolve (presumably) this issue fully at my system. I tried running ncpost 1.1.2, clearing `.{cache,config}/ncspot` and copied `credentials.json` as outputed by [librespot-auth](https://github.com/dspearson/librespot-auth) to `.cache/ncspot/librespot`.

I don’t get a "bad credentials" screen but I cannot play "new" titles, i.e. (I assume) which are not cached. Meaning with my state before, I can only play songs which I played before. And when I clear the cache, I cannot play anything. (Or is my issue different from this one?)

**Comment by blessedbythestorm** [comment]
@Zocker1999NET I'm getting the same. Generating the credentials from librespot-auth then copying them over to '.cache/ncspot/librespot' allows me to continue onto ncspot but I can't play any songs, it will just keep trying to play next until it finds a cached song. At the moment I can't really play any at all.

**Comment by psaumur** [comment]
@Zocker1999NET @blessedbythestorm 
Credentials should be placed here:

~/.var/app/io.github.hrkfdn.ncspot/cache/ncspot/librespot
(or wherever the USER_CACHE_PATH shows when you type 'ncspot info')

If you want to "refresh" ncspot's cache of your library, you can either rename or remove the "files" and "volume" folders in the same directory.

If you want to listen to NEW songs that are not in your playlists, you can using the Search function inside ncspot (F2).

Have been running this on Debian 12 and Linux Mint Edge without any of these issues.

**Comment by blessedbythestorm** [comment]
@psaumur 
In arch `ncspot info` yields `~/.config/ncspot`.

I had this working a few days ago with the `librespot-auth` workaround but then I think the credentials expired and now I can go into ncspot but cant play anything. 
I tried re authenticating to get new credentials but same results.
I tried deleting `.config/ncspot`, `.cache/ncspot` and getting brand new credentials but same result.

https://github.com/user-attachments/assets/336da154-0b81-4da3-9aa9-8d5ffd86ee95

**Comment by Zocker1999NET** [comment]
TL;DR: I have basically the same experience as @blessedbythestorm shows in his last comment [here](https://github.com/hrkfdn/ncspot/issues/1500#issuecomment-2327228665). This also happens on me when trying to play back my saved tracks. Especially after clearing the cache directory, ncspot cannot play any of my 1000+ saved tracks.

<details>
<summary>answer to last message of @psaumur</summary>

(i.e. [this last message](https://github.com/hrkfdn/ncspot/issues/1500#issuecomment-2327015086))

I think we have installed ncspot in different ways. From the path alone, I assume you installed ncspot using flatpak or something alike. I installed ncspot on NixOS from nixpkgs (version 1.1.2 from unstable) or downloaded & compiled it by myself on a Debian machine.

> Credentials should be placed here:
> 
> ~/.var/app/io.github.hrkfdn.ncspot/cache/ncspot/librespot (or wherever the USER_CACHE_PATH shows when you type 'ncspot info')

Hence `ncspot info` in my case returns:
```
USER_CONFIGURATION_PATH /home/zocker/.config/ncspot
USER_CACHE_PATH /home/zocker/.cache/ncspot
USER_RUNTIME_PATH /run/user/1000/ncspot
```

> If you want to "refresh" ncspot's cache of your library, you can either rename or remove the "files" and "volume" folders in the same directory.

As I mentioned before (probably it was not obvious), I already tried running ncspot with cleared cache directory. But then it seems I cannot play any song anymore.

> If you want to listen to NEW songs that are not in your playlists, you can using the Search function inside ncspot (F2).

In my comment, "new" referred to as in "not in cache". I tested ncspot by playing music back

</details>

**Comment by psaumur** [comment]
@Zocker1999NET I did, indeed, install it via Flatpak for both systems.
Have you tried backing up your saved working credentials and reinstalling / recompiling the software?
Does installing it via Flatpak make a difference?

When you use Spotify on the Desktop, do your credential still work or did they force you to update them?
Wonder if Spotify changed something, yet again, and my credentials haven't timed out yet (created Aug. 18th, after it had broke)

**Comment by Denyreal** [comment]
> Not trying to be condescending, but questions like these waste my time and yours. Mine because I could be doing something else; yours because there are plenty of answers online that do a better job than I ever could.

it is the definition of condescending dude.

**Comment by Joachim-42he** [comment]
I have the same problem and have followed the librespot "workaround" to no avail. Before you ask: yes, I put the credentials.json in the right directory. After that, ncspot no longer asked me to login but just dies. I can`t see anything interesting in the debug log (attached) either.
[ncspot_debug.txt](https://github.com/user-attachments/files/16977798/ncspot_debug.txt)

**Comment by hrkfdn** [comment]
https://github.com/librespot-org/librespot/pull/1309#issuecomment-2348066427

With a little bit of luck we may have a working ncspot again next week :)

**Comment by cartesius68** [comment]
Sorry to bother, just waiting to be allowed to run ncspot again. Do we know when logging in will be possible again?

**Comment by yukihooked** [comment]
ngl im about to crashout over this.
I refreshed my oauth and now I can't even use this app anymore, I've tried compiling librespot myself, but the repository doesn't even have any new releases, and I would have to manually fork it.
I just want to use this app again.

**Comment by hrkfdn** [comment]
I have done some prep work in https://github.com/hrkfdn/ncspot/pull/1244, but haven't figured out the correct scopes yet. Also the librespot release isn't published yet, either.

**Comment by cartesius68** [comment]
> I have done some prep work in #1244, but haven't figured out the correct scopes yet. Also the librespot release isn't published yet, either.

Awesome and thanks for your hard work. This app is simply amazing, even nerds like me on FreeBSD are able to play music with it. Now, we're getting better at singing but hopefully everything will return to normal pretty soon. Again, big thanks!

**Comment by hrkfdn** [comment]
There is a first implementation in `main`. Happy to receive reports/fixes. If everything goes well I'll draft a release once librespot is released :)

**Comment by Joachim-42he** [comment]
Hm. Ncspot tells me to login via OAuth2 in my browser and after I have  done that it tells me to return to the terminal but then nothing happens. It could be that my using i3-WM without a DE might lead to problems here.

**Comment by hrkfdn** [comment]
Are you running ncspot on the same computer as your browser? It needs to pass the token to localhost.

**Comment by Joachim-42he** [comment]
Yes I am. I have attached a debug log. 
[ncspot.debug.txt](https://github.com/user-attachments/files/17084008/ncspot.debug.txt)

**Comment by hrkfdn** [comment]
That's odd. Looks like it has received the token and the logs indicate no problem either :confused:

**Comment by BajekekButLost** [comment]
hey, having the same issue here. I installed with winget and I cant log in with a Family plan account. Ik im late to this issue but as I read thru I still couldn't find anything that could fix my issue

**Comment by hrkfdn** [comment]
@BajekekButLost The fix isn't released yet.

**Comment by lucasbeiler** [comment]
> Hm. Ncspot tells me to login via OAuth2 in my browser and after I have done that it tells me to return to the terminal but then nothing happens. It could be that my using i3-WM without a DE might lead to problems here.

Hi, @hrkfdn. On Friday morning I had the same issue as @Joachim-42he after building ncspot in the same way as the official Arch Linux PKGBUILD:
```shell
git clone https://github.com/hrkfdn/ncspot
cargo fetch --locked
export RUSTUP_TOOLCHAIN=stable
export CARGO_TARGET_DIR=target
cargo build --frozen --release --features cover
```
Today, I just built it again, but using `git clone https://github.com/hrkfdn/ncspot && cargo build --release`, and now it works perfectly. It might be due to the latest commit in the main branch since then that may have fixed the problem, or the way I built it on Friday uncovered some bug.

**Comment by psaumur** [comment]
I just used "cargo run" and it worked. Glad it finally resolved for you
using "cargo build"

**Comment by hrkfdn** [comment]
Thanks for the helpful reports, glad it's working :)

**Comment by Drexxl-the-Walrus** [comment]
For me neither cargo run nor starting after cargo build --release works. I just get this error after authentication:

"Failed to exchange code for access token (Server returned error response)"

**Comment by Joachim-42he** [comment]
@hrkfdn It seems that ncspot does not create a cache for the credentials. I have removed the cache folder and rebuilt and rerun ncspot and it still stops after the spotify login. It does create `~/.cache/ncspot/librespot` but that folder remains empty. Can I test somenthing more to help?

**Comment by hrkfdn** [comment]
If the problem also occurs with librespot you may have more luck reporting it there, @Joachim-42he, as we're mostly reusing what they use to perform OAuth2. Raising aware on their end would be much appreciated.

**Comment by hrkfdn** [comment]
@Joachim-42he Hmm, the logs look OK. Is there a backtrace in your cache folder?

**Comment by Joachim-42he** [comment]
@hrkfdn  Yes, there  is
```rust
 0: ncspot::panic::register_backtrace_panic_handler::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::panicking::panic_const::panic_const_div_by_zero
   7: ncspot::application::Application::new
   8: ncspot::main
   9: std::sys::backtrace::__rust_begin_short_backtrace
  10: main
  11: <unknown>
  12: __libc_start_main
  13: _start

panicked at src/ui/cover.rs:51:35:
attempt to divide by zero
```

**Comment by Joachim-42he** [comment]
After simply reading the backtrace and building ncspot without the cover feature, it is running again. Thank you @hrkfdn  Is there a thread where I could be of help debugging this error?

**Comment by rawfootageguy** [comment]
henloo~, it seems I'm the odd Windows user here. I just installed ncspot with scoop (relatively new to this) and ran into the same bad credentials error even after changing my password. For the username field I tried my email and the letter splatter username spotify seems to give you on their back end, but both to no avail. 

As I read this I presume a fix might be out sooner or later? just want to confirm. going to leave ncspot alone for now and come back whenever it might be working again <3

**Comment by undonepotato** [comment]
> henloo~, it seems I'm the odd Windows user here. I just installed ncspot with scoop (relatively new to this) and ran into the same bad credentials error even after changing my password. For the username field I tried my email and the letter splatter username spotify seems to give you on their back end, but both to no avail.
> 
> As I read this I presume a fix might be out sooner or later? just want to confirm. going to leave ncspot alone for now and come back whenever it might be working again <3

@rawfootageguy I've just been reading a lot of things here (since I had to reinstall ncspot and then ran into this issue).  From what I can tell, this issue was fixed in 8b83837 last week and it seems to be working for people (including me). However, this hasn't been made into a release yet, so the only way to use it at this point is by building it from the latest git yourself. You can do this by `git clone`ing the repository and building it with `cargo build --release` (more instructions in [developers.md](https://github.com/hrkfdn/ncspot/blob/main/doc/developers.md#Compiling) and in previous comments; you might have to set an audio backend since you're not on Linux).
Building with the `cover` feature is also causing sudden crashes (which was really frustrating me earlier), so don't do that if you plan to build right now since that's probably a bug.

Alternatively, you could just wait for the next release/scoop to add it, where this will be fixed and you can use your normal installation method.

However you get ncspot working eventually, make sure to use your letter-splatter username -- email doesn't work.

Hope this helps and correct me if I'm wrong!

**Comment by hrkfdn** [comment]
I'm reopening this for visibility until we can publish a new release as the issue reports keep coming.

The bottom line is: this should be fixed in `main`. You'll either have to compile the `main` branch or wait until the librespot release (one of ncspot's dependencies) is out, so that I can publish as well.

**Comment by ifneq** [comment]
I've been struggling with compiling for the past few hours, can you make a build guide for Windows or MacOS? Whenever I try to build on MacOS or Windows, I get a ``cc.exe`` or ``linker.exe`` error stating:
LINK : fatal error LNK1181: cannot open input file ':libpulse-simple-0.dll.lib'

**Comment by ocntscha** [comment]
Same credentials problem her with my Spotify premium family account after installing on Fedora40 Linux through its built in app installer. 

The work around of creating a credentials.json file with librespot-auth utility and placing it in ~/.var/app/io.github.hrkfdn.ncspot/cache/ncspot/librespot/ was amazingly easy since I already have official the Spotify client installed and there is a statically compiled binary of librespot-auth in the releases which was super simple and worked perfectly

**Comment by Gh0stjaw** [comment]
Had the same issue in Void Linux, with the package installed from `xbps-install ncspot`.

Did this to make it work..

First just removed the xbps package 
`sudo xbps-remove ncspot`
then:
`git clone https://github.com/hrkfdn/ncspot`
`cargo install --path to/where/you/put/your/git/clone`
(only other issue but non related, I were missing openssl-devel, just installed it through `sudo xbps-install openssl-devel`).

At first run it gave me a long url (The OAuth which were missing from the initial install) to just open with a browser, then its all good to go :-)

**Comment by hrkfdn** [comment]
ncspot 1.2.0 is out! This should be fixed now.

**Comment by Tachi107** [comment]
@hrkfdn great news! I'd personally keep this pinned for a little longer so that people using an older version are more likely to see this and avoid reporting a useless issue

**Comment by shawny43** [comment]
Update your flatpak and add a tool to get quicker access please! I don't want to pass again trough flatpak running commands for each boot up.

**Comment by kalyanoliveira** [comment]
I'm using version 1.2.2. This is still broken. Installed using Nix.

**Comment by indiedeveloperGPU** [comment]
Solution for invalid credentials was TLS fingerprint
